### PR TITLE
fix(segment-analytic) : triggering events with analytic

### PR DESCRIPTION
**Pull request by VanjaPin**
https://applicaster.monday.com/boards/1615228456/views/71924469/pulses/7123126451

### CODE REVIEW

File: src/source/segment_analytics/SegmentAnalytics.brs [hunk 1]
```diff
@@ -62,6 +56,67 @@ function SegmentAnalytics(config as object, port as object) as object
...
+
+  if config.apiHost = invalid or (type(config.apiHost) <> "roString" and type(config.apiHost) <> "String") then
+    updatedConfig.apiHost = "https://api.segment.io"
+  end if
+
+  if config.settingsApiHost = invalid or (type(config.settingsApiHost) <> "roString" and type(config.settingsApiHost) <> "String") then
+    updatedConfig.settingsApiHost = "https://cdn-settings.segment.com"
+  end if
+
+  if config.factories = invalid or type(config.factories) <> "roAssociativeArray" then
```
Comment (georgejecook) [inline src/source/segment_analytics/SegmentAnalytics.brs:87]: See next comment, we should use asAA here.

File: src/source/segment_analytics/SegmentAnalytics.brs [hunk 2]
```diff
@@ -62,6 +56,67 @@ function SegmentAnalytics(config as object, port as object) as object
...
+
+  if config.queueSize = invalid or (type(config.queueSize) <> "roInteger" and type(config.queueSize) <> "roInt") or config.queueSize < 1 then
+    updatedConfig.queueSize = 1
+  end if
+
+  if config.retryLimit = invalid or (type(config.retryLimit) <> "roInteger" and type(config.retryLimit) <> "roInt") or config.retryLimit < -1 then
+    updatedConfig.retryLimit = 1
+  end if
+
+  if config.requestPoolSize = invalid or (type(config.requestPoolSize) <> "roInteger" and type(config.requestPoolSize) <> "roInt") or config.requestPoolSize < 1 then
```
Comment (georgejecook) [inline src/source/segment_analytics/SegmentAnalytics.brs:75]: we should use `if asInt(config.requestPoolSize, invalid) = invalid`

File: src/source/segment_analytics/SegmentAnalytics.brs [hunk 3]
```diff
@@ -62,6 +56,67 @@ function SegmentAnalytics(config as object, port as object) as object
...
 
+'Create configuration with default values for any values not explicitly included
+function _SegmentAnalytics_configWithDefaultValues(config as object) as object
+  updatedConfig = config
+
+  if config.debug = invalid or (config.debug <> false and config.debug <> true) then
+    updatedConfig.debug = false
+  end if
+
+  if config.queueSize = invalid or (type(config.queueSize) <> "roInteger" and type(config.queueSize) <> "roInt") or config.queueSize < 1 then
```
Comment (georgejecook) [inline src/source/segment_analytics/SegmentAnalytics.brs:67]: we should use `if asInt(config.requestPoolSize, invalid) = invalid`

File: src/source/segment_analytics/SegmentAnalytics.brs [hunk 4]
```diff
@@ -62,6 +56,67 @@ function SegmentAnalytics(config as object, port as object) as object
   }
 end function
 
+'Create configuration with default values for any values not explicitly included
+function _SegmentAnalytics_configWithDefaultValues(config as object) as object
+  updatedConfig = config
+
+  if config.debug = invalid or (config.debug <> false and config.debug <> true) then
```
Comment (georgejecook) [inline src/source/segment_analytics/SegmentAnalytics.brs:63]: we should use `if asBool(config.requestPoolSize, invalid) = invalid`

File: src/source/segment_analytics/SegmentAnalytics.brs [hunk 5]
```diff
@@ -62,6 +56,67 @@ function SegmentAnalytics(config as object, port as object) as object
...
+
+  if config.defaultSettings = invalid or type(config.defaultSettings) <> "roAssociativeArray" then
+    updatedConfig.defaultSettings = {}
+  end if
+
+  if config.defaultSettings.integrations = invalid or type(config.defaultSettings.integrations) <> "roAssociativeArray" then
+    updatedConfig.defaultSettings.integrations = {}
+  end if
+
+  if config.defaultSettings.integrations["Segment.io"] = invalid or type(config.defaultSettings.integrations["Segment.io"]) <> "roAssociativeArray" then
```
Comment (georgejecook) [inline src/source/segment_analytics/SegmentAnalytics.brs:103]: use  `if asAA(config.defaultSettings.integrations["Segment.io"], invalid) = invalid`

File: src/source/segment_analytics/SegmentAnalytics.brs [hunk 6]
```diff
@@ -282,22 +337,22 @@ function _SegmentAnalytics_checkValidId(data as dynamic) as boolean
...
-
-  if not hasUserId and not hasAnonId then
-    callType = "unknown"
-    if not data.type = invalid and (type(data.type) = "roString" or type(data.type) = "String")
-      callType = data.type
-    end if
-    m._log("No user or anonymous id found in [" + callType + "] call", "ERROR")
-    return false
-  end if
+ if data.userId <> invalid and (type(data.userId) = "roString" or type(data.userId) = "String") then
```
Comment (georgejecook) [inline src/source/segment_analytics/SegmentAnalytics.brs:340]: use the `if asString(xxx, invalid) = invalid` check

File: src/source/segment_analytics/SegmentAnalytics.brs [hunk 7]
```diff
@@ -282,22 +337,22 @@ function _SegmentAnalytics_checkValidId(data as dynamic) as boolean
...
+   hasUserId = data.userId.len() > 0
+ end if
+
+ if data.anonymousId <> invalid and (type(data.anonymousId) <> "roString" or type(data.anonymousId) <> "String") then
```
Comment (georgejecook) [inline src/source/segment_analytics/SegmentAnalytics.brs:344]: use the `if asString(xxx, invalid) = invalid` check

File: src/source/segment_analytics/SegmentAnalytics.brs [hunk 8]
```diff
@@ -282,22 +337,22 @@ function _SegmentAnalytics_checkValidId(data as dynamic) as boolean
...
+   hasAnonId = data.anonymousId.len() > 0
+ end if
+
+ if not hasUserId and not hasAnonId then
+   callType = "unknown"
+   if not data.type = invalid and (type(data.type) = "roString" or type(data.type) = "String")
```
Comment (georgejecook) [inline src/source/segment_analytics/SegmentAnalytics.brs:350]: use the `if asString(xxx, invalid) = invalid` check

File: src/source/segment_analytics/SegmentAnalyticsPlugin.bs [hunk 9]
```diff
@@ -14,8 +14,8 @@ class SegmentAnalyticsPlugin extends zapp.analytics.BaseAnalyticsPlugin
   private isIdentifyingOnLogout
   private library
   private segmentTask
-  private defaultOptions
-  private defaultProperties
+  private defaultOptions = {}
+  private defaultProperties = {}
```
Comment (georgejecook) [inline src/source/segment_analytics/SegmentAnalyticsPlugin.bs:18]: This file has unit tests, they should be updated.

**Comment by VanjaPin** [comment]
fixme: asString and asAa, etc do not work here

**Review by georgejecook — APPROVED** [review]

**Pull request merged**